### PR TITLE
Add upstream changes and clean up where possible

### PR DESCRIPTION
**Pull request by dan-blanchard**
This is very much a work in progress at the moment, and I'm just creating the PR to make it easier for me to keep track of Travis results.

I have a few goals for this branch:

1.  Pull in changes from Mozilla's upstream code.  There aren't as many as I had initially expected but there are some.
2.  Improve PEP8 compliance all over the place.  The previous maintainers tried to keep variable names identical to the C code, presumably to ease the comparison with the Mozilla code, but we're going to be diverging from upstream after pulling in the changes mentioned in 1.  Basically, Mozilla seems very likely to abandon their character encoding detector in the near future and [switch to using ICU](https://bugzilla.mozilla.org/show_bug.cgi?id=724533), but ICU doesn't support all of the codecs we currently do, because it is more web-focused.  If our goal here is to be a truly universal character encoding detector, we'll need to go our own way in the future in that respect.
3.  Make the unit tests pass, or at the very least make it obvious that the tests are actually failing (instead of ignoring the failures like our current Travis build does).

So far, I've done a little bit of point 1 and updated the Travis testing setup to use nose and report test coverage via Coveralls.

### CODE REVIEW

**Comment by dan-blanchard** [comment]
@sigmavirus24 I know this is a total bear of a PR at this point, but most of the changes are purely cosmetic.  The main substantive changes are:

1.  PEP8 compliance as much as possible (especially with variable names).
2.  Implemented filter_english_letters (#46)
3.  Switched to using enums (via enum34 and Python 3.4 standard library) instead of constants in a few places.
4.  Added a bunch of comments to UniversalDetector (I'll try to add more to others later).
5.  Added the same language filtering support present in the upstream Mozilla code (which appears to be the only substantive change they made other than commenting out the Hungarian detectors).

I'd like to propose a more substantial refactoring of things to make it simpler to retrain some of the models, which were never updated after they were established ~15 years ago, but I'll file a separate issue about that.

If you get a chance to look this over, I'd appreciate your thoughts.

**Comment by dan-blanchard** [comment]
I should mention that there are also fewer test failures (we're down to 21), because I fixed some issues where we were being tested for encodings that we didn't even support.

**Comment by sigmavirus24** [comment]
Looks like I have my work cut out for me. :+1: *starts reviewing*

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,11 @@
 PKG-INFO
 /dist
 /build
+.idea/.name
```
Comment (sigmavirus24) [inline .gitignore]: This belongs in a [global gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore).

File: .travis.yml [hunk 2]
```diff
@@ -3,11 +3,24 @@ sudo: false
...
   - 2.7
-  - 3.2
   - 3.3
+  - 3.4
+  - pypy
 
-script: python test.py
+install:
+  - travis_retry pip install python-coveralls nose-cov
+  - python setup.py install
```
Comment (sigmavirus24) [inline .travis.yml]: What about `pip install .`?

File: NOTES.rst [hunk 3]
```diff
@@ -1,4 +1,122 @@
...
+byte sequence.  The reason this extra step is necessary is that the frequency
+rank table is language-specific (and not encoding-specific).
+
+
+What's where
+============
+
+
+Bigram files
+------------
```
Comment (sigmavirus24) [inline NOTES.rst]: Aren't there supposed to be new-lines between the `---`s and the list?

File: NOTES.rst [hunk 4]
```diff
@@ -1,4 +1,122 @@
...
+- euckrprober.py
+- euctwprober.py
+- gb2312prober.py
+- mbcharsetprober.py
+- mbcsgroupprober.py
+- sjisprober.py
+
+Misc files
+----------
+- __init__.py (currently has detect function in it)
```
Comment (sigmavirus24) [inline NOTES.rst]: Will this display funky? Should we escape it like:

```rst
- ``__init__.py``
```

File: chardet/big5freq.py [hunk 5]
```diff
@@ -381,545 +381,6 @@
  938,3941, 553,2680, 116,5783,3942,3667,5784,3545,2681,2783,3438,3344,2820,5785, # 5328
 3668,2943,4160,1747,2944,2983,5786,5787, 207,5788,4809,5789,4810,2521,5790,3033, # 5344
  890,3669,3943,5791,1878,3798,3439,5792,2186,2358,3440,1652,5793,5794,5795, 941, # 5360
-2299, 208,3546,4161,2020, 330,4438,3944,2906,2499,3799,4439,4811,5796,5797,5798, # 5376  #last 512
```
Comment (sigmavirus24) [inline chardet/big5freq.py]: Where did all of this go?

File: chardet/big5freq.py [hunk 6]
```diff
@@ -381,545 +381,6 @@
...
-13888,13889,13890,13891,13892,13893,13894,13895,13896,13897,13898,13899,13900,13901,13902,13903, #13904
-13904,13905,13906,13907,13908,13909,13910,13911,13912,13913,13914,13915,13916,13917,13918,13919, #13920
-13920,13921,13922,13923,13924,13925,13926,13927,13928,13929,13930,13931,13932,13933,13934,13935, #13936
-13936,13937,13938,13939,13940,13941,13942,13943,13944,13945,13946,13947,13948,13949,13950,13951, #13952
-13952,13953,13954,13955,13956,13957,13958,13959,13960,13961,13962,13963,13964,13965,13966,13967, #13968
-13968,13969,13970,13971,13972) #13973
-
+2299, 208,3546,4161,2020, 330,4438,3944,2906,2499,3799,4439,4811,5796,5797,5798, # 5376
+)
 # flake8: noqa
```
Comment (sigmavirus24) [inline chardet/big5freq.py]: Can we remove that?

File: chardet/charsetprober.py [hunk 7]
```diff
@@ -26,70 +26,120 @@
...
+            # If the last character in the word is a marker, replace it with a
             # space as markers shouldn't affect our analysis (they are used
             # similarly across all languages and may thus have similar
-            # frequencies)
+            # frequencies).
             last_char = word[-1:]
-            last_char = last_char if last_char.isalpha() or \
-                last_char >= b'\x80' else b' '
+            last_char = last_char if (last_char.isalpha() or
+                                      last_char >= b'\x80') else b' '
```
Comment (sigmavirus24) [inline chardet/charsetprober.py]: This looks terrible as a ternary. Let's just make this a plain and simple `if`/`else` block

File: chardet/charsetprober.py [hunk 8]
```diff
@@ -26,70 +26,120 @@
...
+        This filter can be applied to all scripts which contain both English
+        characters and extended ASCII characters, but is currently only used by
+        ``Latin1Prober``.
+        """
+        filtered = BytesIO()
+        in_tag = False
+        prev = 0
+
+        for curr in range(len(buf)):
+            buf_char = buf[curr:curr + 1]
```
Comment (sigmavirus24) [inline chardet/charsetprober.py]: For people unfamiliar with why this is necessary we should put a short comment on the fact this is necessary for byte"strings" on Py3 more than anywhere else.

File: chardet/compat.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
 ######################## BEGIN LICENSE BLOCK ########################
 # Contributor(s):
-#   Ian Cordasco - port to Python
```
Comment (sigmavirus24) [inline chardet/compat.py]: I must have written this but it certainly wasn't part of a port to python

File: chardet/enums.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+class ProbingState(IntEnum):
+    """
+    This enum represents the different states a prober can be in.
+    """
+    detecting = 0
+    found_it = 1
+    not_me = 2
+
+
+class SMState(IntEnum):
```
Comment (sigmavirus24) [inline chardet/enums.py]: I understand that `StateMachineState` sound terribly redundant, but can we at least name it something like `MachineState`? `SMState` isn't much better than some of the old variable names that didnt' communicate anything that well.

File: chardet/escsm.py [hunk 11]
```diff
@@ -223,20 +223,20 @@
...
-                    'stateTable': ISO2022KR_st,
-                    'charLenTable': ISO2022KRCharLenTable,
-                    'name': "ISO-2022-KR"}
+ISO2022KR_SM_MODEL = {'class_table': ISO2022KR_CLS,
+                      'class_factor': 6,
+                      'state_table': ISO2022KR_ST,
+                      'char_len_table': ISO2022KR_CHAR_LEN_TABLE,
+                      'name': "ISO-2022-KR"}
 
 # flake8: noqa
```
Comment (sigmavirus24) [inline chardet/escsm.py]: Should probably remove this now too

File: chardet/gb2312prober.py [hunk 12]
```diff
@@ -28,14 +28,15 @@
...
-from .mbcssm import GB2312SMModel
+from .mbcssm import GB2312_SM_MODEL
 
 class GB2312Prober(MultiByteCharSetProber):
     def __init__(self):
-        MultiByteCharSetProber.__init__(self)
-        self._mCodingSM = CodingStateMachine(GB2312SMModel)
-        self._mDistributionAnalyzer = GB2312DistributionAnalysis()
+        super(GB2312Prober, self).__init__()
+        self.coding_sm = CodingStateMachine(GB2312_SM_MODEL)
```
Comment (sigmavirus24) [inline chardet/gb2312prober.py]: Maybe this is the first I've noticed this, but why make this public now?

**Comment by sigmavirus24** [comment]
So I have a few thoughts that I've left as comments. I noticed you selectively made one attribute of the probers public. Given the defined API for the library, why not make everything public? Or leave everything private as it was? I'm not sure I understand the point of it. Seems like unnecessary churn.

Further, if we're going to care about style, we should add it as a Travis CI job to at least catch regressions. Or we could add landscape.io. If we want to do the former, I'll send a PR to this branch with a set-up I use on pretty much every other project that makes it significantly easier.

File: NOTES.rst (hunk 4 above)
Comment (dan-blanchard) [inline NOTES.rst]: Yes, but I really was just updating that file to make some notes about things for myself. I didn't think anyone else was likely to read it.  I'll fix it either way though. :)

File: NOTES.rst (hunk 3 above)
Comment (dan-blanchard) [inline NOTES.rst]: The blank lines are optional according to the RST spec.

File: NOTES.rst (hunk 4 above)
Comment (dan-blanchard) [inline NOTES.rst]: Actually, I just checked [here](https://github.com/chardet/chardet/blob/feature/upstream-changes-and-overhaul/NOTES.rst) and it renders fine as is.

Comment (sigmavirus24) [inline NOTES.rst]: GitHub's rendering (of reStructuredText) is historically broken. Relying on that has bitten me in the past.

File: chardet/big5freq.py (hunk 5 above)
Comment (dan-blanchard) [inline chardet/big5freq.py]: It was all commented out in the upstream code, so it just seemed like a waste of space/memory to keep them.  The comment seems to indicate that nothing after that point occurs is one of the 512 most frequent characters for Big5.

File: .travis.yml (hunk 2 above)
Comment (dan-blanchard) [inline .travis.yml]: Any reason to prefer that over `python setup.py install` if we're using setuptools?

Comment (sigmavirus24) [inline .travis.yml]: Is that not how almost every one of our users will be installing chardet?

Comment (dan-blanchard) [inline .travis.yml]: Hmm... good point. :+1:

File: NOTES.rst (hunk 4 above)
Comment (dan-blanchard) [inline NOTES.rst]: Okay, I'll make all of the filenames have surrounding backticks then for consistent formatting.

File: chardet/big5freq.py (hunk 6 above)
Comment (dan-blanchard) [inline chardet/big5freq.py]: Sure. I don't personally use flake8 (I prefer pylint), so I'm fine with taking all of those out.

**Comment by dan-blanchard** [comment]
> Further, if we're going to care about style, we should add it as a Travis CI job to at least catch regressions. Or we could add landscape.io. If we want to do the former, I'll send a PR to this branch with a set-up I use on pretty much every other project that makes it significantly easier.

Wow, I hadn't seen landscape.io before. That seems pretty nice.  I'm fine with using that if you are.

**Comment by dan-blanchard** [comment]
> So I have a few thoughts that I've left as comments. I noticed you selectively made one attribute of the probers public. Given the defined API for the library, why not make everything public? Or leave everything private as it was? I'm not sure I understand the point of it. Seems like unnecessary churn.

That was a mishap of all the massive find-and-replacing I was doing. That said, I'd honestly prefer to make everything public that doesn't currently need/use a property.

File: chardet/compat.py (hunk 9 above)
Comment (dan-blanchard) [inline chardet/compat.py]: Yeah, that's what I figured. :smile:

**Comment by dan-blanchard** [comment]
@sigmavirus24 I believe I've addressed all of your comments at this point.

**Comment by dan-blanchard** [comment]
I've gone ahead and disabled our Hungarian probers (like the Mozilla authors did). This fixes all but one of our unit test failures.  The remaining failure is for a Greek XML file where we misidentify ISO-8859-7 as Windows-1253, which leads to a one character difference in the decoded output.  Without retraining our models, I'm not sure how we could actually fix that one.

The difference is:
```diff
-  δραστήριου τετραπληγικού. Η εργασία είναι στην Αθήνα, στον Άγιο Νικόλαο (Αχαρνών). Τα καθήκοντά αφορούν στην υποστήριξη των υλικών καθημερινών αναγκών, την 
?                                                             ^

+  δραστήριου τετραπληγικού. Η εργασία είναι στην Αθήνα, στον ¶γιο Νικόλαο (Αχαρνών). Τα καθήκοντά αφορούν στην υποστήριξη των υλικών καθημερινών αναγκών, την 
?                                                             ^
```

**Comment by dan-blanchard** [comment]
I'm going to merge this because I believe I've addressed all of the objections @sigmavirus24 and we have much bigger future direction things to think about at the moment.

**Comment by sigmavirus24** [comment]
So I just realized, @dan-blanchard that you added a reliance on `enum34` which is going to pose problems for requests. Any chance we can revert that work since it isn't necessary or a public API change?

**Comment by dan-blanchard** [comment]
@sigmavirus24 why is the enum34 reliance an issue for requests?  Since requests is probably the main way that chardet gets used, we can change things for it if necessary, but IMHO the code is a lot cleaner using enums.

**Comment by sigmavirus24** [comment]
Because Kenneth insists on vendoring requests' dependencies. Which means dependencies with dependencies need to either vendor them or they're axed. Maybe once we can convince Kenneth to stop vendoring it won't be a problem, but this will prevent requests users from getting any of the bug fixes introduced in this mega pull request.

Disclaimer: [I think Enums are superior too.](https://review.openstack.org/#/c/149738/)

**Comment by dan-blanchard** [comment]
@sigmavirus24 I just made our enums inherit from `object` instead of `IntEnum` because we weren't actually using any of the fancier features.  That should eliminate the problem. 02edb35

**Pull request merged**